### PR TITLE
[FSDP] Fix input grad propagation when using param mixed precision

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90921 [FSDP] Fix input grad propagation when using param mixed precision**
* #90858 [FSDP][Easy] ufmt files
* #90840 [FSDP][BE] Remove `_module_to_handles`, `HandleConfig`; use term "fqn"; clarify docs

For parameter mixed precision, we cast the inputs to the low precision parameter dtype. If the input has tensors that require gradient, then we must cast them in place in order for them to receive a gradient. The cast should be tracked by autograd (e.g. with `grad_fn` equal to `ToCopyBackward0`). This removes the `torch.no_grad` context when calling `_apply_to_tensors`.

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -456,12 +456,13 @@ def _cast_fp_inputs_to_dtype(
     def cast_fn(x: torch.Tensor) -> torch.Tensor:
         if not torch.is_floating_point(x) or x.dtype == dtype:
             return x
-        y = x.to(dtype)
-        # Explicitly copy over `requires_grad` since this runs inside
-        # `torch.no_grad()`
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:461]: I think this comment about `_cast_fp_inputs_to_dtype` running in `no_grad` is outdated given my check just now.

**Review by mrshenli — APPROVED** [review]
LGTM

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -456,15 +456,9 @@ def _cast_fp_inputs_to_dtype(
...
-        if x.is_leaf:
-            y.requires_grad = x.requires_grad
-        return y
-
-    with torch.no_grad():
-        return (_apply_to_tensors(cast_fn, args), _apply_to_tensors(cast_fn, kwargs))
+        return x.to(dtype)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:459]: hmm, I am a bit concerned about unforseen BC issues, but also our testing surface is quite solid. Did we consider doing this only for inputs that require grad to err on the safe side?

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 3]
```diff
@@ -718,6 +718,50 @@ def never_wrap_policy(*args, **kwargs):
...
+            mixed_precision=mixed_precision,
+            device_id=torch.cuda.current_device(),
+            use_orig_params=use_orig_params,
+        )
+        # Use an input with dtype not equal to the mixed precision
+        # `param_dtype` so that it gets cast
+        x_float = torch.randn(32, 1024, device="cuda", dtype=torch.float32)
+        x_float.requires_grad_()
+        fsdp_model(x_float).sum().backward()
+        self.assertTrue(x_float.grad is not None)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:763]: nit: should we also assert on the dtype?

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 4]
```diff
@@ -718,6 +718,50 @@ def never_wrap_policy(*args, **kwargs):
...
+        fsdp_model = FSDP(
+            model,
+            sharding_strategy=sharding_strategy,
+            mixed_precision=mixed_precision,
+            device_id=torch.cuda.current_device(),
+            use_orig_params=use_orig_params,
+        )
+        # Use an input with dtype not equal to the mixed precision
+        # `param_dtype` so that it gets cast
+        x_float = torch.randn(32, 1024, device="cuda", dtype=torch.float32)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:760]: nit: you can pass `requires_grad=True` and avoid the set below.

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for the fix!

File: test/distributed/fsdp/test_fsdp_mixed_precision.py (hunk 3 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_mixed_precision.py:763]: Good point.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:459]: I think if `x` does not require gradient, then `x.to(dtype)` will not be tracked by autograd. if `x` does require gradient, then `x.to(dtype)` will be tracked by autograd. This should be exactly the behavior we want. In other words, the casing should be handled naturally already if I understand correctly.
